docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+# ADR 004: Coroutine-Based Effect System for Consensus
+
+## Changelog
+
+* 18-03-2025: Initial version
+
+## Status
+
+Accepted
+
+## Context
+
+The Malachite consensus engine needs to interact with its environment (network, storage, cryptography, application logic) at specific points during execution.
+
+Traditional approaches to handling these interactions include:
+
+1. **Callback-based designs** - Provide callback interfaces that must be implemented by users
+2. **Trait-based polymorphism** - Define traits that users implement to provide required functionality
+3. **Message-passing architectures** - Define protocols for communication between consensus and external components
+
+We needed a design that would:
+- Maintain a clear separation between the consensus algorithm and its environment
+- Keep the consensus code linear and readable despite external interactions
```

Comment:
Linear here is something like "identical to the pseudo-code"?

Reply:
Kind of, I meant devoid of callbacks or indirections. Like the difference between structuring code as a state machine where the flow of the program is not immediately obvious from reading the code compared to a program that simply calls other functions and act on their return values.